ambari-server/src/main/java/org/apache/ambari/server/state/UpgradeHelper.java

```diff
@@ -464,6 +423,56 @@ public UpgradePack suggestUpgradePack(String clusterName,
             continue;
           }
 
+          HostsType hostsType = mhr.getMasterAndHosts(service.serviceName, component);
+
+          // only worry about adding future commands if this is a start/restart task
+          boolean taskIsRestartOrStart = functionName == null || functionName == Type.START
+              || functionName == Type.RESTART;
+
+          // see if this component has an add component task which will indicate
+          // we need to dynamically schedule some more tasks by predicting where
+          // the components will be installed
+          String serviceAndComponentHash = service.serviceName + "/" + component;
+          if (taskIsRestartOrStart && addedComponentsDuringUpgrade.containsKey(serviceAndComponentHash)) {
```

Comment:
How about moving this block to a separate method?

Reply:
I thought about it, but I'd have to pass in a ton of parameters and it wouldn't be reused anywhere.